proposed/2021/TargetFrameworkAndPlatformsFilters.md

```diff
@@ -0,0 +1,196 @@
+
+1. The **generation of .NET** such as .NET Framework, .NET Core, .NET (5, 6, and later), and Any.
+2. The **supported platforms** for the generation such as .NET Standard, Windows, macOS, Linux, Android, iOS, and more.
+3. The **version number** of the generation being used such as .NET 5, .NET Core 3.1, .NET Framework 4.8, and more.
+
+This proposal only accounts for the **majority of packages on NuGet.org**. There are many unique TFMs in the .NET ecosystem, but for the sake of creating a concept that will scale with the future of .NET, many unsupported & low usage TFMs are not considered. Thus this proposal only focuses on the [most common target frameworks](https://docs.microsoft.com/dotnet/standard/frameworks#latest-versions) & omits considering TFMs with fewer than 5,000 packages.
+
+#### Generations of .NET
+
+There are many generations of .NET. For the sake of this proposal, we will focus on defining the three major generations:
+
+- .NET Framework:
+  - `net, net11, net20, net35, net40, net403, net45, net451, net452, net46, net461, net462, net463, net47, net471, net472, net48`
+- .NET Core:
+  - `netcoreapp1.0, netcoreapp1.1, netcoreapp2.0, netcoreapp2.1, netcoreapp2.2, netcoreapp3.0, netcoreapp3.1`
+- .NET:
+  - `net5.0, net6.0`
+- Any:
+  - .NET Framework, .NET Core, .NET
+
+![](meta/resources/TargetFrameworkAndPlatformFilters/GenerationsOfNET.png)
+
+#### Supported Platforms
+
+With .NET Core introducing support for cross-platform scenarios, we will define the following as supported platforms:
+
+- .NET Standard:
+  - `netstandard, netstandard1.0, netstandard1.1, netstandard1.2, netstandard1.3, netstandard1.4, netstandard1.5, netstandard1.6, netstandard2.0, netstandard2.1`
+- Windows:
+  - .NET Framework
+  - .NET Core
+  - .NET 5+
+    - `net5.0-windows, net6.0-windows`
+  - .NET Standard
+- macOS:
+  - .NET Core
+  - .NET 5+
+    - `net6.0-macos`
+  - .NET Standard
+- Linux:
+  - .NET Core
+  - .NET 5+
+  - .NET Standard
+- Android:
+  - .NET 6+
+    - `net6.0-android`
+  - .NET Standard
+- iOS:
+  - .NET 6+
+    - `net6.0-ios`
+  - .NET Standard
+- macCatalyst:
+  - .NET 6+
+    - `net6.0-maccatalyst`
+  - .NET Standard
+- tvOS:
+  - .NET 6+
+    - `net6.0-tvos`
+  - .NET Standard
+- Tizen:
+  - .NET 6+
+    - `net6.0-tizen`
+  - .NET Standard
+
+**Note:** .NET Standard is considered a special case that supports majority of these platforms depending on the [version of .NET standard](https://docs.microsoft.com/dotnet/standard/net-standard#net-implementation-support). If the framework precedence of a target framework supports .NET Standard, it's respective versions supported will be implicitly included.
+
+![](meta/resources/TargetFrameworkAndPlatformFilters/SupportedPlatforms.png)
+
+#### Version Numbers
+
+There are many unique version numbers for all generations of .NET & supported platforms. The filtering experience should be smart enough to create a version list based on the current selection. Here are a few scenarios:
```

Comment:
When I saw the screenshot my first reaction was "why would anyone want to list all packages that has a version 2.0".

I now see it's about TFMs, but it still raises some questions for me:
1) Are the selected TFMs what a _package_ targets, or what TFMs _I_ target and hence a nuget must support.
E.g. if my app targets net462 and netcoreapp2.0, a nuget targeting netstandard2.0 would be compatible.

2) Regarding the "The filtering experience should be smart enough to create a version list based on the current selection". I'm wondering about the "smart" part.
Say my app multi targets .net framework 4.7.2 and .net core 2.0.
Would selecting `.NET Framework` and `.NET Core` and then selecting `4.7.2` and `2.0` from the version selector imply that the combined TFMs selected are `.NET Framework 2.0`, `.NET Framework 4.7.2` and `.NET Core 2.0`?

Reply:
1. It would be what a package provides assets for. At least for now. So in your case, if you filtered for `.NET Framework` -> `Windows` -> `4.6.2`, you would see packages providing assets for `net462` or are compatible with `net462`. https://nugettools.azurewebsites.net/6.0.0-preview.3/framework-precedence?framework=net462&exludedIdentifiers=
2. Yes, that is right. You would get packages matching `.NET Standard 2.0` as well in this case as its implicitly included. One note is that the current proposal uses `Any` for all generations(.NET Framework and .NET Core and .NET)  & only allows individual selection outside of that(.NET Framework or .NET Core or .NET).

Reply:
I'm still/more confused about what your response 1 is saying.

> what a package provides assets for

My understanding is that this means if a package only has `lib/netstandard2.0/my.dll`, and I search `framework:net462`, this package will **not** be in the search results. It will only return packages that have `lib/net462/*` (or from the other assets types, not just lib).

> or are compatible in precedence with net462

So it is returning all packages that are compatible with projects that are targeting net462, directly contracting my understanding of the above?

Reply:
For a `net462` scenario, the precedence includes `netstandard1.3 -> netstandard2.0`. .NET Standard is implicitly included in search results & thus would be included.

If we allow the ability to disable the implicit `.NET Standard` functionality, then only packages that have `net462` and prior `net` version assets would show up. 

There's probably a further discussion on where do we draw the line on precedence for older/out of support TFMs that are still "compatible". I think maybe that line to draw is on the framework itself & netstandard.

Reply:
Resolving as this has a path forward. Added a note to `unresolved questions` for now.